File: android/src/main/java/expo/modules/inappmessaging/ExpoInAppMessagingModule.kt
package expo.modules.inappmessaging

import android.util.Log
import com.google.firebase.inappmessaging.FirebaseInAppMessaging
import com.google.firebase.inappmessaging.FirebaseInAppMessagingDisplayCallbacks
import com.google.firebase.inappmessaging.model.BannerMessage
import com.google.firebase.inappmessaging.model.CardMessage
import com.google.firebase.inappmessaging.model.ImageOnlyMessage
import com.google.firebase.inappmessaging.model.InAppMessage
import com.google.firebase.inappmessaging.model.Action
import com.google.firebase.inappmessaging.model.MessageType
import com.google.firebase.inappmessaging.model.ModalMessage
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoInAppMessagingModule : Module() {

  private var currentCallbacks: FirebaseInAppMessagingDisplayCallbacks? = null
  private var currentActionUrl: String = ""

  override fun definition() = ModuleDefinition {
    Name("ExpoInAppMessaging")

    Events("onMessage")

    OnCreate {
      FirebaseInAppMessaging.getInstance().setMessageDisplayComponent { message, callbacks ->
        currentCallbacks = callbacks
        currentActionUrl = extractActionUrl(message)

        val payload = buildPayload(message)
        sendEvent("onMessage", payload)
        callbacks.impressionDetected()
      }
    }

    Function("setMessagesSuppressed") { suppressed: Boolean ->
      FirebaseInAppMessaging.getInstance().setMessagesSuppressed(suppressed)
    }

    Function("triggerEvent") { eventName: String ->
      FirebaseInAppMessaging.getInstance().triggerEvent(eventName)
    }

    AsyncFunction("getPendingMessage") {
      return@AsyncFunction null
    }


    AsyncFunction("logClick") {
      try {
        val action = Action.builder().setActionUrl(currentActionUrl).build()
        currentCallbacks?.messageClicked(action)
        currentCallbacks = null
      } catch (e: Exception) {
        Log.e("ExpoInAppMessaging", "Failed to log click: ${e.message}")
      }
    }


    AsyncFunction("logDismiss") {
      try {
        currentCallbacks?.messageDismissed(
          FirebaseInAppMessagingDisplayCallbacks.InAppMessagingDismissType.UNKNOWN_DISMISS_TYPE
        )
        currentCallbacks = null
      } catch (e: Exception) {
        Log.e("ExpoInAppMessaging", "Failed to log dismiss: ${e.message}")
      }
    }
  }


  private fun extractActionUrl(message: InAppMessage): String {
    return when (message.messageType) {
      MessageType.CARD -> (message as? CardMessage)?.primaryAction?.actionUrl ?: ""
      MessageType.MODAL -> (message as? ModalMessage)?.action?.actionUrl ?: ""
      MessageType.BANNER -> (message as? BannerMessage)?.action?.actionUrl ?: ""
      MessageType.IMAGE_ONLY -> (message as? ImageOnlyMessage)?.action?.actionUrl ?: ""
      else -> ""
    }
  }

  private fun buildPayload(message: InAppMessage): Map<String, Any?> {
    val map = mutableMapOf<String, Any?>()

    map["type"] = message.messageType?.name ?: "UNKNOWN"
    map["campaignId"] = message.campaignMetadata?.campaignId ?: ""
    map["campaignName"] = message.campaignMetadata?.campaignName ?: ""
    map["isTestMessage"] = message.campaignMetadata?.isTestMessage ?: false
    map["data"] = message.data ?: emptyMap<String, String>()
    map["title"] = ""
    map["body"] = ""
    map["imageUrl"] = ""
    map["actionUrl"] = currentActionUrl

    when (message.messageType) {
      MessageType.CARD -> {
        val card = message as CardMessage
        map["title"] = card.title.text ?: ""
        map["body"] = card.body?.text ?: ""
        map["imageUrl"] = card.portraitImageData?.imageUrl ?: ""
      }
      MessageType.MODAL -> {
        val modal = message as ModalMessage
        map["title"] = modal.title.text ?: ""
        map["body"] = modal.body?.text ?: ""
        map["imageUrl"] = modal.imageData?.imageUrl ?: ""
      }
      MessageType.BANNER -> {
        val banner = message as BannerMessage
        map["title"] = banner.title.text ?: ""
        map["body"] = banner.body?.text ?: ""
        map["imageUrl"] = banner.imageData?.imageUrl ?: ""
      }
      MessageType.IMAGE_ONLY -> {
        val img = message as ImageOnlyMessage
        map["imageUrl"] = img.imageData?.imageUrl ?: ""
      }
      else -> {}
    }

    return map
  }
}

File: ios/ExpoInAppMessagingModule.swift
import ExpoModulesCore
import FirebaseInAppMessaging

public class ExpoInAppMessagingModule: Module {
  private static var pendingMessage: [String: Any]?

  public func definition() -> ModuleDefinition {
    Name("ExpoInAppMessaging")

    Events("onMessage")


    OnCreate {
      if let delegate = ExpoInAppMessagingLifecycleDelegate.instance {
        delegate.onMessage = { [weak self] message in
          guard let self = self else { return }

          let payload = self.parseMessage(message)
          Self.pendingMessage = payload

          delegate.currentDisplayDelegate?.impressionDetected?(for: message)

          self.sendEvent("onMessage", payload)
        }
      }
    }

 
    AsyncFunction("getPendingMessage") { () -> [String: Any]? in
      let message = Self.pendingMessage
      Self.pendingMessage = nil
      return message
    }

    Function("setMessagesSuppressed") { (suppressed: Bool) in
      InAppMessaging.inAppMessaging().messageDisplaySuppressed = suppressed
    }

    Function("triggerEvent") { (eventName: String) in
      InAppMessaging.inAppMessaging().triggerEvent(eventName)
    }

    AsyncFunction("logClick") { () -> Void in
      guard
        let delegate = ExpoInAppMessagingLifecycleDelegate.instance,
        let displayDelegate = delegate.currentDisplayDelegate,
        let message = delegate.currentMessage
      else { return }

      let actionURL = self.getActionURL(message)
      let action = InAppMessagingAction(
        actionText: nil,
        actionURL: actionURL.isEmpty ? nil : URL(string: actionURL)
      )
      displayDelegate.messageClicked?(message, with: action)
      delegate.currentMessage = nil
      delegate.currentDisplayDelegate = nil
    }


    AsyncFunction("logDismiss") { () -> Void in
      guard
        let delegate = ExpoInAppMessagingLifecycleDelegate.instance,
        let displayDelegate = delegate.currentDisplayDelegate,
        let message = delegate.currentMessage
      else { return }

      displayDelegate.messageDismissed?(message, dismissType: .typeUserTapClose)
      delegate.currentMessage = nil
      delegate.currentDisplayDelegate = nil
    }
  }

    
  private func parseMessage(_ message: InAppMessagingDisplayMessage) -> [String: Any] {
    var payload: [String: Any] = [
      "type": messageTypeName(message),
      "campaignId": message.campaignInfo.messageID,
      "campaignName": message.campaignInfo.campaignName ?? "",
      "data": message.appData ?? [:],
      "title": "",
      "body": "",
      "imageUrl": "",
      "actionUrl": ""
    ]

    if let msg = message as? InAppMessagingBannerDisplay {
      payload["title"] = msg.title
      payload["body"] = msg.bodyText ?? ""
      payload["imageUrl"] = msg.imageData?.imageURL ?? ""
      payload["actionUrl"] = msg.actionURL?.absoluteString ?? ""
    } else if let msg = message as? InAppMessagingModalDisplay {
      payload["title"] = msg.title
      payload["body"] = msg.bodyText ?? ""
      payload["imageUrl"] = msg.imageData?.imageURL ?? ""
      payload["actionUrl"] = msg.actionURL?.absoluteString ?? ""
    } else if let msg = message as? InAppMessagingCardDisplay {
      payload["title"] = msg.title
      payload["body"] = msg.body ?? ""
      payload["imageUrl"] = msg.portraitImageData.imageURL
      payload["actionUrl"] = msg.primaryActionURL?.absoluteString ?? ""
    } else if let msg = message as? InAppMessagingImageOnlyDisplay {
      payload["imageUrl"] = msg.imageData.imageURL
      payload["actionUrl"] = msg.actionURL?.absoluteString ?? ""
    }

    return payload
  }

  private func messageTypeName(_ message: InAppMessagingDisplayMessage) -> String {
    if message is InAppMessagingBannerDisplay { return "BANNER" }
    if message is InAppMessagingModalDisplay { return "MODAL" }
    if message is InAppMessagingCardDisplay { return "CARD" }
    if message is InAppMessagingImageOnlyDisplay { return "IMAGE_ONLY" }
    return "UNKNOWN"
  }

  private func getActionURL(_ message: InAppMessagingDisplayMessage) -> String {
    if let msg = message as? InAppMessagingBannerDisplay {
      return msg.actionURL?.absoluteString ?? ""
    }
    if let msg = message as? InAppMessagingModalDisplay {
      return msg.actionURL?.absoluteString ?? ""
    }
    if let msg = message as? InAppMessagingCardDisplay {
      return msg.primaryActionURL?.absoluteString ?? ""
    }
    if let msg = message as? InAppMessagingImageOnlyDisplay {
      return msg.actionURL?.absoluteString ?? ""
    }
    return ""
  }
}

